localstack/services/kms/provider.py

```diff
@@ -133,89 +163,238 @@ def __init__(self, message=None):
+        if not request.get("KeyId"):
+            raise ValidationError("Required input parameter KeyId not specified")
+        store = self._get_store(context)
+        # KeyId can potentially hold one of multiple different types of key identifiers. Here we find a key no
+        # matter which type of id is used.
+        key_id = store.get_canonical_key_id(request.get("KeyId"))
+
+        grant_id = request.get("GrantId")
+        if grant_id:
+            if grant_id not in store.grants:
+                raise InvalidGrantIdException()
+            return ListGrantsResponse(Grants=[store.grants[grant_id].metadata])
+
+        matching_grants = []
+        grantee_principal = request.get("GranteePrincipal")
+        for grant in store.grants.values():
+            # KeyId is a mandatory field of ListGrants request, so is going to be present.
+            if grant.metadata["KeyId"] != key_id:
+                continue
+            # GranteePrincipal is a mandatory field for CreateGrant, should be in grants. But it is an optional field
+            # for ListGrants, so might not be there.
+            if grantee_principal and grant.metadata["GranteePrincipal"] != grantee_principal:
+                continue
+            matching_grants.append(grant.metadata)
+
+        grants_list = PaginatedList(matching_grants)
+        page, next_token = grants_list.get_page(
+            lambda grant_data: grant_data.get("GrantId"),
+            next_token=request.get("Marker"),
+            page_size=request.get("Limit", 50),
+        )
+        kwargs = {"NextMarker": next_token, "Truncated": True} if next_token else {}
+
+        return ListGrantsResponse(Grants=page, **kwargs)
+
+    # Honestly, this is a mess in AWS KMS. Hashtag "do we follow specifications that are a pain to customers or do we
+    # diverge from AWS and make the life of our customers easier?"
+    #
+    # Both RetireGrant and RevokeGrant operations delete a grant. The differences between them are described here:
+    # https://docs.aws.amazon.com/kms/latest/developerguide/grant-manage.html#grant-delete
+    # Essentially:
+    # - Permissions to RevokeGrant are controlled through IAM policies or through key policies, while permissions to
+    # RetireGrant are controlled by settings inside the grant itself.
+    # - A grant to be retired can be specified by its GrantToken or its GrantId/KeyId pair. While revoking grants can
+    # only be done with a GrantId/KeyId pair.
+    # - For RevokeGrant, KeyId can be either an actual key ID, or an ARN of that key. While for RetireGrant only key
+    # ARN is accepted as a KeyId.
+    #
+    # We currently do not model permissions for retirement and revocation of grants. At least not in KMS,
+    # maybe IAM in LocalStack has some modelling though. We also accept both key IDs and key ARNs for both
+    # operations. So apart from RevokeGrant not accepting GrantToken parameter, we treat these two operations the same.
+    def _delete_grant(
```

Comment:
Can also be static as it doesn't use any provider instance data

Reply:
Done.